### PR TITLE
Fix widget visibility tracking algorithm

### LINKED ISSUE #1437: Frame visibility in sitemap ignored
PR #1174 looks to be included in 2.7.10-beta, so it may have caused this, or it isn't included and may fix this issue.

### Actual behaviour
Sitemaps with items that should be hidden are always visible. Basic UI displays them properly.

### Expected behaviour
Sitemaps with items that should be hidden should be hidden.

### Steps to reproduce
1. Add Item1 to a sitemap inside a frame with visibility based on Item2.
2. Toggle Item2 and notice visibility of Item1 does not change.

Here is a snip from the sitemap where I see this occurring...
```
    Frame visibility=[Mode=="Morning"] {
        Switch item=Wakeup_Alarm
        Switch item=Wakeup_Alarm_Snooze
    }
```

#### Client
* Android version: 2.7.10-beta

#### Server
* Server version: S1627

### PR DESCRIPTION
**Pull request by maniac103**
Closes #1437

### CODE REVIEW

**Comment by openhab-5iver** [comment]
Thank you!

File: mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt [hunk 1]
```diff
@@ -202,8 +203,29 @@ class WidgetAdapter(
...
+        // If visibility of a container with at least one child changes, refresh the whole list to make sure
+        // the child visibility is also updated. Otherwise it's sufficient to update the single widget only.
+        if (oldWidget.visibility != widget.visibility && items.any { w -> w.parentId == widget.id }) {
+            notifyDataSetChanged()
+        } else {
+            notifyItemChanged(position)
+        }
+    }
+
+    private tailrec fun isWidgetIncludingAllParentsVisible(widget : Widget): Boolean {
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt]: AS claims `tailrec` isn't needed here.

Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt]: Where do you see this? I don't see any notice like that in the code editor, during build or in the 'analyze code' output.

Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt]: This is a lint inspection that is run during editing code.

Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt]: OK. Looks like one would need to help the compiler here:

```diff
diff --git a/mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt b/mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt
index 1a5609c1..ce681cda 100644
--- a/mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt
+++ b/mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt
@@ -220,8 +220,8 @@ class WidgetAdapter(
         if (!widget.visibility) {
             return false
         }
-        val parent = widget.parentId?.let { id -> widgetsById[id] }
-        return parent == null || isWidgetIncludingAllParentsVisible(parent)
+        val parent = widget.parentId?.let { id -> widgetsById[id] } ?: return true
+        return isWidgetIncludingAllParentsVisible(parent)
     }
 
     private fun getItemViewType(widget: Widget): Int {
```

I've checked in the bytecode output that it indeed currently doesn't do tail recursion optimization.

**Pull request merged**